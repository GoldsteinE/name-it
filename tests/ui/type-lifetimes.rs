use futures::{pin_mut, FutureExt as _};

use name_it::name_it;

struct Wrapper<'a>(&'a str);

async fn foo() {}

#[name_it(Test)]
async fn has_lifetime(x: Wrapper<'_>) -> String {
    foo().await;
    x.0.to_string()
}

fn main() {
    let x = String::from("test");
    let y = has_lifetime(Wrapper(x.as_str()));
    pin_mut!(y);
    drop(x);
    y.now_or_never().unwrap();
}
