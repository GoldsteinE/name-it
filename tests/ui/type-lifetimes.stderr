error[E0505]: cannot move out of `x` because it is borrowed
  --> tests/ui/type-lifetimes.rs:19:10
   |
17 |     let y = has_lifetime(Wrapper(x.as_str()));
   |                                  ---------- borrow of `x` occurs here
18 |     pin_mut!(y);
19 |     drop(x);
   |          ^ move out of `x` occurs here
20 |     y.now_or_never().unwrap();
21 | }
   | - borrow might be used here, when `y` is dropped and runs the destructor for type `Named<_has_lifetime_impl::Test<'_>>`
